src/data/qpf/multivariate/functor.lean

```diff
@@ -0,0 +1,669 @@
+/-
+Copyright (c) 2018 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jeremy Avigad, Mario Carneiro
+
+Tuples of types, and their categorical structure.
+
+Features:
+
+`typevec n`   : n-tuples of types
+`α ⟹ β`      : n-tuples of maps
+`f ⊚ g`       : composition
+`mvfunctor n` : the type class of multivariate functors
+`f <$$> x`    : notation for map
+
+Also, support functions for operating with n-tuples of types, such as:
+
+`append1 α β`    : append type `β` to n-tuple `α` to obtain an (n+1)-tuple
+`drop α`         : drops the last element of an (n+1)-tuple
+`last α`         : returns the last element of an (n+1)-tuple
+`append_fun f g` : appends a function g to an n-tuple of functions
+`drop_fun f`     : drops the last function from an n+1-tuple
+`last_fun f`     : returns the last function of a tuple.
+
+Since e.g. `append1 α.drop α.last` is propositionally equal to `α` but not definitionally equal
+to it, we need support functions and lemmas to mediate between constructions.
+-/
+import data.fin logic.function.basic
+
+/-
+n-tuples of types, as a category
+-/
+
+universes u v w
+
+inductive fin' : ℕ → Type
```

Comment:
We already have `fin2` (in a file about diophantine equations).  This should be moved to `data/fin2.lean` and add the missing theorems from here.